Mathlib/RingTheory/Polynomial/Hilbert.lean

```diff
@@ -0,0 +1,133 @@
+  `d`, if `F` is of characteristic `0`, then `RatFunc.Polynomial.hilbert p d : F[X]` is the
+  polynomial whose value at `n` equals the coefficient of `Xⁿ` in the power series expansion of
+  `p/(1 - X)ᵈ`.
+
+## TODO
+
+* Hilbert polynomials of finitely generated graded modules over Noetherian rings.
+-/
+
+open Nat Polynomial PowerSeries
+
+variable (F : Type*) [Field F]
+
+namespace RatFunc
+
+namespace Polynomial
+
+/--
+For any field `F` and natural numbers `d` and `k`, `RatFunc.Polynomial.preHilbert F d k`
+is defined as `(d.factorial : F)⁻¹ • ((ascPochhammer F d).comp (X - (C (k : F)) + 1))`.
+This is the most basic form of Hilbert polynomials. `RatFunc.Polynomial.preHilbert ℚ d 0`
+is exactly the Hilbert polynomial of the polynomial ring `ℚ[X_0,...,X_d]` viewed as
+a graded module over itself. In fact, `RatFunc.Polynomial.preHilbert F d k` is the
+same as `RatFunc.Polynomial.hilbert ((X : F[X]) ^ k) (d + 1)` for any field `F` and
+`d k : ℕ` (see the lemma `RatFunc.Polynomial.hilbert_X_pow_succ`). See also the lemma
+`RatFunc.Polynomial.preHilbert_eq_choose_sub_add`, which states that if `CharZero F`,
+then for any `d k n : ℕ` with `k ≤ n`, `(RatFunc.Polynomial.preHilbert F d k).eval (n : F)`
+equals `(n - k + d).choose d`.
+-/
+noncomputable def preHilbert (d k : ℕ) : F[X] :=
+  (d.factorial : F)⁻¹ • ((ascPochhammer F d).comp (Polynomial.X - (C (k : F)) + 1))
+
+lemma preHilbert_eq_choose_sub_add [CharZero F] (d : ℕ) {k n : ℕ} (hkn : k ≤ n):
+    (preHilbert F d k).eval (n : F) = (n - k + d).choose d := by
+  have : ((d ! : ℕ) : F) ≠ 0 := by norm_cast; positivity
+  calc
+  _ = (↑d !)⁻¹ * eval (↑(n - k + 1)) (ascPochhammer F d) := by simp [cast_sub hkn, preHilbert]
+  _ = (n - k + d).choose d := by
+    rw [ascPochhammer_nat_eq_natCast_ascFactorial];
+    field_simp [ascFactorial_eq_factorial_mul_choose]
+
+variable {F}
+
+/--
+`RatFunc.Polynomial.hilbert p 0 = 0`; for any `d : ℕ`, `RatFunc.Polynomial.hilbert p (d + 1)`
+is defined as `∑ i in p.support, (p.coeff i) • RatFunc.Polynomial.preHilbert F d i`. If
+`M` is a graded module whose Poincaré series can be written as `p(X)/(1 - X)ᵈ` for some
+`p : ℚ[X]` with integer coefficients, then `RatFunc.Polynomial.hilbert p d` is the Hilbert
+polynomial of `M`. See also `RatFunc.Polynomial.coeff_mul_invOneSubPow_eq_hilbert_eval`,
+which says that `PowerSeries.coeff F n (p * (PowerSeries.invOneSubPow F d))` is equal to
+`(RatFunc.Polynomial.hilbert p d).eval (n : F)` for any large enough `n : ℕ`.
+-/
+noncomputable def hilbert (p : F[X]) : (d : ℕ) → F[X]
+  | 0 => 0
+  | d + 1 => ∑ i in p.support, (p.coeff i) • preHilbert F d i
+
+variable (F) in
+lemma hilbert_zero_nat (d : ℕ) : hilbert (0 : F[X]) d = 0 := by
+  delta hilbert; induction d with
+  | zero => simp only
+  | succ d _ => simp only [coeff_zero, zero_smul, Finset.sum_const_zero]
+
+lemma hilbert_poly_zero (p : F[X]) : hilbert p 0 = 0 := by
+  delta hilbert; simp
```

Comment:
```suggestion
lemma hilbert_poly_zero (p : F[X]) : hilbert p 0 = 0 := rfl
```
Can you add `hilbert_poly_succ` and the fact that it is `F`-linear?

Reply:
I have added `hilbert_poly_succ`. I will add the fact that it is `F`-linear in another PR.